src/main/java/org/opentripplanner/graph_builder/module/osm/naming/SidewalkNamer.java

```diff
@@ -0,0 +1,275 @@
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.geotools.api.referencing.FactoryException;
+import org.geotools.api.referencing.crs.CoordinateReferenceSystem;
+import org.geotools.api.referencing.operation.MathTransform;
+import org.geotools.api.referencing.operation.TransformException;
+import org.geotools.geometry.jts.JTS;
+import org.geotools.referencing.CRS;
+import org.geotools.referencing.crs.DefaultGeographicCRS;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.MultiLineString;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.operation.buffer.BufferParameters;
+import org.locationtech.jts.operation.distance.DistanceOp;
+import org.opentripplanner.framework.geometry.GeometryUtils;
+import org.opentripplanner.framework.geometry.HashGridSpatialIndex;
+import org.opentripplanner.framework.geometry.SphericalDistanceLibrary;
+import org.opentripplanner.framework.i18n.I18NString;
+import org.opentripplanner.framework.lang.DoubleUtils;
+import org.opentripplanner.framework.logging.ProgressTracker;
+import org.opentripplanner.graph_builder.services.osm.EdgeNamer;
+import org.opentripplanner.openstreetmap.model.OSMWithTags;
+import org.opentripplanner.street.model.edge.StreetEdge;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A namer that assigns names of nearby streets to sidewalks if they meet certain
+ * geometric similarity criteria.
+ * <p>
+ * The algorithm works as follows:
+ *  - for each sidewalk we look up (named) street edges nearby
+ *  - group those edges into groups where each edge has the same name
+ *  - draw a flat-capped buffer around the sidewalk, like this: https://tinyurl.com/4fpe882h
+ *  - check how much of a named edge group is inside the buffer
+ *  - remove those groups which are below MIN_PERCENT_IN_BUFFER
+ *  - take the group that has the highest percentage (as a proportion of the sidewalk length) inside
+ *    the buffer and apply its name to the sidewalk.
+ * <p>
+ * This works very well for OSM data where the sidewalk runs a parallel to the street and at each
+ * intersection the sidewalk is also split. It doesn't work well for sidewalks that go around
+ * the corner, like https://www.openstreetmap.org/way/1059101564. These cases are, however, detected
+ * by the above algorithm and the sidewalk name remains the same.
+ */
+public class SidewalkNamer implements EdgeNamer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SidewalkNamer.class);
+  private static final int MAX_DISTANCE_TO_SIDEWALK = 50;
+  private static final double MIN_PERCENT_IN_BUFFER = .85;
+  private static final int BUFFER_METERS = 25;
+
+  private HashGridSpatialIndex<EdgeOnLevel> streetEdges = new HashGridSpatialIndex<>();
```

Comment:
I think you right: deciding when to index is a rather muddled affair and quite ad-hoc. I would definitely welcome and improvement here.

I think a solution would be to establish a relationship between graph and index where the graph notifies the index of insertions and deletions so that it can be updated as a stream rather than always through a full re-index.

Reply:
In this case though we want to not index any edge but only those which have an explicit name and which are not highway links. Sure with some refactoring these could also be filtered out during comparison within this class.

Reply:
Thanks for the clarification. Refactoring all that indexing throughout the graph build process is out of scope for this PR, and indexes are already being built several times, so I guess one more time here is reasonable. Changing that indexing system could be a task for another time. Like a few other things in the PR, I just mentioned this index mainly because the inefficiency/cost may grow with the size of the graph.